src/ops/tag.rs

```diff
@@ -65,13 +65,32 @@ pub enum OpTag {
 impl OpTag {
     /// Returns true if the tag is more general than the given tag.
     #[inline]
-    pub fn contains(self, other: OpTag) -> bool {
-        self == other || other.parent_tags().iter().any(|&tag| self.contains(tag))
+    pub const fn contains(self, other: OpTag) -> bool {
+        // We cannot call iter().any() or even do for loops in const fn yet.
+        // So we have to write this ugly code.
+        if self.eq(other) {
```

Comment:
I don't really understand this logic. The comment says we're looking for something more general, but the name implies we're querying whether a given optag can contain another as a child node. If it's the latter then, say, a leaf node can't contain another leaf node

Reply:
Ah, we are not talking about hierarchies here.

Tags are sets of operations (hence `TAG::Any` and `TAG::None`). So we are looking at set inclusion with `contains`.

The method `parent_tags` returns the immediate successors in the set inclusion order, so we traverse the order using that (perhaps it should be renamed...).

Reply:
`contains` -> `is_superset`
`parent_tags` -> `immediate_supersets`

hopefully more readable now